protocol/contracts/templegold/TempleGoldVesting.sol

```diff
@@ -0,0 +1,444 @@
+     * @param _revokeTimestamp Time to revoke
+     */
+    function revokeVesting(bytes32 _vestingId, uint32 _revokeTimestamp) external override onlyElevatedAccess {
+        if (_revokeTimestamp < block.timestamp) { revert CommonEventsAndErrors.InvalidParam(); }
+        VestingSchedule storage _schedule = schedules[_vestingId];
+        if (!_vestingIdExists(_vestingId)) { revert CommonEventsAndErrors.InvalidParam(); }
+        if (_revokeTimestamp >= _schedule.duration + _schedule.start) { revert CommonEventsAndErrors.InvalidParam(); }
+        if (_schedule.revoked) { revert VestingRevoked(); }
+        uint256 _vestedAmount = _calculateReleasableAmount(_schedule);
+        if (_vestedAmount > 0) {
+            _release(_schedule, _vestedAmount);
+            emit Released(_vestingId, _schedule.recipient, _vestedAmount);
+        }
+        uint256 unreleased = _schedule.amount - _schedule.distributed;
+        // update new revoke duration
+        if (_revokeTimestamp > block.timestamp) {
+            uint256 maxAmount = _calculateTotalVestedAt(_schedule, _revokeTimestamp);
+            revokedMaxAmounts[_vestingId] = maxAmount;
+            uint256 remainingAmount = maxAmount - _calculateTotalVestedAt(_schedule, uint32(block.timestamp));
+            unreleased -= remainingAmount;
+        }
+        totalVested -= unreleased;
+        _schedule.revoked = true;
+        _activeVestingIds.remove(_vestingId);
+        emit Revoked(_vestingId, _schedule.recipient, unreleased, totalVested);
+    }
+
+    /**
+     * @notice Recover ERC20 token
+     * @param _token Token address
+     * @param _to Recipient address
+     * @param _amount Amount to recover
+     */
+    function recoverToken(address _token, address _to, uint256 _amount) external override onlyElevatedAccess {
+        emit CommonEventsAndErrors.TokenRecovered(_to, _token, _amount);
+        IERC20(_token).safeTransfer(_to, _amount);
+    }
+
+    /**
+     * @dev Computes the next vesting schedule identifier for a given account address.
+     */
+    function computeNextVestingScheduleIdForHolder(
+        address _recipient
+    ) public view override returns (bytes32) {
+        return
+            computeVestingScheduleIdForAddressAndIndex(
+                _recipient,
+                _holdersVestingCount[_recipient]
+            );
+    }
+
+    /**
+     * @notice Get vesting schedules count by recipient
+     * @param _recipient Recipient address
+     * @return Count
+     */
+    function getVestingSchedulesCountByRecipient(
+        address _recipient
+    ) external view override returns (uint256) {
+        return _holdersVestingCount[_recipient];
+    }
+
+    /**
+     * @notice Get vesting Id at index from list of vesting Ids
+     * @param _index Index
+     * @return id Bytes32 Id
+     */
+    function getVestingIdAtIndex(
+        uint256 _index
+    ) external view override returns (bytes32 id) {
+        if (_index >= _vestingIds.length()) { return id; }
```

Comment:
Is this check needed? It looks like it will just return zero anyway within EnumerableSet if the index isn't found?

```
    function _at(Set storage set, uint256 index) private view returns (bytes32) {
        return set._values[index];
    }
```